chained/selector.go

```diff
@@ -0,0 +1,214 @@
+
+	// maxDialerFails is the max consecutive failed dials/requests before moving to the next dialer
+	maxDialerFails = 4
+)
+
+// Dialer provides the ability to dial a proxy
+type Dialer interface {
+
+	// DialProxy dials the proxy but does not yet dial the origin.
+	DialProxy(ctx context.Context) (net.Conn, error)
+
+	// SupportsAddr indicates whether this Dialer supports the given addr. If it does not, the
+	// balancer will not attempt to dial that addr with this Dialer.
+	SupportsAddr(network, addr string) bool
+
+	// DialContext dials out to the given origin. failedUpstream indicates whether
+	// this was an upstream error (as opposed to errors connecting to the proxy).
+	DialContext(ctx context.Context, network, addr string) (conn net.Conn, failedUpstream bool, err error)
+
+	// Name returns the name for this Dialer
+	Name() string
+
+	// Label returns a label for this Dialer (includes Name plus more).
+	Label() string
+
+	// JustifiedLabel is like Label() but with elements justified for line-by
+	// -line display.
+	JustifiedLabel() string
+
+	// Location returns the country code, country name and city name of the
+	// dialer, in this order.
+	Location() (string, string, string)
+
+	// Protocol returns a string representation of the protocol used by this
+	// Dialer.
+	Protocol() string
+
+	// Addr returns the address for this Dialer
+	Addr() string
+
+	// Trusted indicates whether or not this dialer is trusted
+	Trusted() bool
+
+	// NumPreconnecting returns the number of pending preconnect requests.
+	NumPreconnecting() int
+
+	// NumPreconnected returns the number of preconnected connections.
+	NumPreconnected() int
+
+	// MarkFailure marks a dial failure on this dialer.
+	MarkFailure()
+
+	// EstRTT provides a round trip delay time estimate, similar to how RTT is
+	// estimated in TCP (https://tools.ietf.org/html/rfc6298)
+	EstRTT() time.Duration
+
+	// EstBandwidth provides the estimated bandwidth in Mbps
+	EstBandwidth() float64
+
+	// EstSuccessRate returns the estimated success rate dialing this dialer.
+	EstSuccessRate() float64
+
+	// Attempts returns the total number of dial attempts
+	Attempts() int64
+
+	// Successes returns the total number of dial successes
+	Successes() int64
+
+	// ConsecSuccesses returns the number of consecutive dial successes
+	ConsecSuccesses() int64
+
+	// Failures returns the total number of dial failures
+	Failures() int64
+
+	// ConsecFailures returns the number of consecutive dial failures
+	ConsecFailures() int64
+
+	// Succeeding indicates whether or not this dialer is currently good to use
+	Succeeding() bool
+
+	// DataSent returns total bytes of application data sent to connections
+	// created via this dialer.
+	DataSent() uint64
+	// DataRecv returns total bytes of application data received from
+	// connections created via this dialer.
+	DataRecv() uint64
+
+	// Stop stops background processing for this Dialer.
+	Stop()
+
+	WriteStats(w io.Writer)
+}
+
+type Selector struct {
```

Comment:
Nit: this name is too general to be meaningful IMO.  Maybe `LoopDialer`?  Something like that?